src/org/pentaho/reporting/platform/plugin/cache/DeleteOldOnAccessCache.java

```diff
@@ -81,18 +92,32 @@ private void cleanUp() {
     final long currentTimeMillis = System.currentTimeMillis();
     final ICacheBackend backend = getBackend();
     //Check all timestamps
-    for ( final String key : backend.listKeys( Collections.singletonList( SEGMENT ) ) ) {
-      if ( key.matches( ".*" + TIMESTAMP ) ) {
-        final List<String> compositeKey = Arrays.asList( SEGMENT, key );
-        final Long timestamp = (Long) backend.read( compositeKey );
-        if ( currentTimeMillis - timestamp > millisToLive ) {
-          backend.purge( compositeKey );
-          compositeKey.set( 1, compositeKey.get( 1 ).replace( TIMESTAMP, "" ) );
-          backend.purge( compositeKey );
-          logger.debug( "Purged long-term cache: " + key );
+    for ( final String userFolder : backend.listKeys( Collections.singletonList( SEGMENT ) ) ) {
+      for ( final String key : backend.listKeys( Arrays.asList( SEGMENT, userFolder ) ) ) {
+        if ( key.matches( ".*" + TIMESTAMP ) ) {
+          final List<String> compositeKey = Arrays.asList( SEGMENT, userFolder, key );
+          final Long timestamp = (Long) backend.read( compositeKey );
+          if ( currentTimeMillis - timestamp > millisToLive ) {
+            backend.purge( compositeKey );
+            compositeKey.set( NAME_INDEX, compositeKey.get( NAME_INDEX ).replace( TIMESTAMP, "" ) );
+            backend.purge( compositeKey );
+            logger.debug( "Purged long-term cache: " + key );
+          }
         }
       }
     }
     logger.debug( "Finished periodical cache eviction" );
   }
+
+
+  private String createKey( final String key ) {
+    try {
+      final MessageDigest md = MessageDigest.getInstance( "SHA-256" );
+      md.update( key.getBytes() );
+      final byte[] digest = md.digest();
+      return new String( Base64.encode( digest ) );
+    } catch ( final Exception b ) {
+      return null;
+    }
```

Comment:
Returning <null> might not be a good idea. Throw an error and let the request fail (as we have no meaningful way of recovering here). 

In addition to that, a <null> element as a segment in the key for the storage backend is not handled and would cause an NPE. However, I cannot see a safe case where <null> should ever be allowed as segment id.